services/worker/src/worker/main.py

```diff
@@ -1,53 +1,90 @@
-# SPDX-License-Identifier: Apache-2.0
-# Copyright 2022 The HuggingFace Authors.
+import json
+import logging
+import os
+import sys
+import tempfile
+import time
+from typing import Optional
 
+from filelock import FileLock
 from libcommon.log import init_logging
-from libcommon.processing_graph import ProcessingGraph
-from libcommon.resources import CacheMongoResource, QueueMongoResource
-from libcommon.storage import init_assets_dir
+from libcommon.queue import Job, Status, get_datetime
+from libcommon.resources import QueueMongoResource
+from mirakuru import OutputExecutor
 
+from worker import start_worker_loop
 from worker.config import AppConfig
-from worker.job_runner_factory import JobRunnerFactory
-from worker.loop import Loop
-from worker.resources import LibrariesResource
+from worker.loop import WorkerState
+
+WORKER_STATE_FILE_NAME = "worker_state.json"
+START_WORKER_LOOP_PATH = start_worker_loop.__file__
+
+
+class WorkerExecutor:
+    def __init__(self, app_config: AppConfig) -> None:
+        self.app_config = app_config
+
+    def _create_worker_loop_executor(self) -> OutputExecutor:
+        banner = self.app_config.worker.state_path
+        if not banner:
+            raise ValueError("Failed to create the executor because WORKER_STATE_PATH is missing.")
+        start_worker_loop_command = [
+            sys.executable,
+            START_WORKER_LOOP_PATH,
+            "--print-worker-state-path",
+        ]
+        return OutputExecutor(start_worker_loop_command, banner, timeout=10)
+
+    def start(self) -> None:
+        worker_loop_executor = self._create_worker_loop_executor()
+        worker_loop_executor.start()  # blocking until the banner is printed
+        logging.info("Starting heartbeat.")
+        while worker_loop_executor.running():
+            self.heartbeat()
+            time.sleep(self.app_config.worker.heartbeat_time_interval_seconds)
+        worker_loop_executor.stop()
+
+    def get_state(self) -> WorkerState:
+        worker_state_path = self.app_config.worker.state_path
+        if not worker_state_path:
+            raise ValueError("Failed to get worker state because WORKER_STATE_PATH is missing.")
+        if os.path.exists(worker_state_path):
+            with FileLock(worker_state_path + ".lock"):
```

Comment:
same as above: we could set the filename before, instead of appending .lock everytime (not a performance issue, just simpler code)